README.md

```diff
@@ -1,526 +1,917 @@
+  [MY_META]: 123,
+}
+```
 
-#### `@on` with a mixin
+This object would then be assigned to another object representing all of the metadata on the class _or_ class prototype, depending on its placement. Static class elements are placed on one object, and non-static class elements are placed on another. So for instance, this example:
 
 ```js
-class MyElement extends WithActions(HTMLElement) {
-  @on('click') clickHandler() { }
+const MY_META = Symbol();
+
+function myMeta(value, context) {
+  context.defineMetadata("my-meta", true);
+  context.defineMetadata(MY_META, true);
+}
+
+@myMeta
+class C {
+  @myMeta foo = 123
+  @myMeta bar() {}
+
+  @myMeta static baz = 123;
+  @myMeta static qux() {}
 }
 ```
 
-This decorator could be defined as follows:
+Would produce the following two metadata objects:
 
 ```js
-const handler = Symbol("handler");
-function on(eventName)
-  return (method, context) => {
-    context.metadata = {[handler]: eventName};
-    return method;
-  }
+let staticMeta = {
+  constructor: { "my-meta": true, [MY_META]: true },
+  baz: { "my-meta": true, [MY_META]: true },
+  qux: { "my-meta": true, [MY_META]: true },
 }
 
-class MetadataLookupCache {
-  #map = new WeakMap();
-  #name;
-  constructor(name) { this.#name = name; }
-  get(newTarget) {
-    let data = this.#map.get(newTarget);
-    if (data === undefined) {
-      data = [];
-      let klass = newTarget;
-      while (klass !== null && !(this.#name in klass)) {
-        for (const [name, {[this.#name]: eventName}]
-             of Object.entries(klass[Symbol.metadata].instance.methods)) {
-          if (eventName !== undefined) {
-            data.push({name, eventName});
-          }
-        }
-        klass = klass.__proto__;
-      }
-      this.#map.set(newTarget, data)
-    }
-    return data;
-  }
+let nonStaticMeta = {
+  foo: { "my-meta": true, [MY_META]: true },
+  bar: { "my-meta": true, [MY_META]: true },
 }
+```
 
-let handlersMap = new MetadataLookupCache(handler);
+Note that metadata defined by a class decorator is assigned to the static `constructor` key. This is because `constructor` is a reserved name within class definitions, so it cannot conflict with another class element with the same name.
 
-function WithActions(superclass) {
-  return class C extends superclass {
-    constructor(...args) {
-      super(...args);
-      let handlers = handlersMap.get(new.target, C);
-      for (const {name, eventName} of handlers) {
-        this.addEventListener(eventName, this[name].bind(this));
-      }
-    }
-  }
+These metadata objects are then exposed via the `Symbol.metadata` property on the class (for static metadata) and the class prototype (for non-static metadata). So the above example is roughly equivalent, when executed, to:
+
+```js
+C[Symbol.metadata] = {
+  constructor: { "my-meta": true, [MY_META]: true },
+  baz: { "my-meta": true, [MY_META]: true },
+  qux: { "my-meta": true, [MY_META]: true },
+};
+
+C.prototype[Symbol.metadata] = {
```

Comment:
My thinking was that they would not be, given this is not an extremely common pattern, and `init-method`s make it even less necessary. Decorators on elements of the same name would be coalesced, and if your decorator needed to do something specific for the field or the method, it could include the value it decorated in the metadata itself (e.g. using `kind`).

I think @ljharb's point about the collision between private fields and dynamic public fields is more of a problem though, since those are actually _different_ names, one is just the "spelling" of the private field but isn't at all related to the public field. I think that means that we'll need to introduce namespacing for private values no matter what, so we could consider namespacing different types of elements as well. I was mainly worried about a combinatorial explosion there (private fields, public fields, private methods, public methods, etc...) but maybe that is not avoidable.